### DISCUSSION TITLE
Layer listeners

### DISCUSSION THREAD

**Discussion by ChrisEll87** [mapsforge/mapsforge#1433] [Q&A] [answered]
I have a GUI (in Swing) that shows all the layer (groups) in order to (de-)activate them on the map. For this usecase I need to know when a layer was added to the map. Of course, there is an `onAdd()` and `onRemove()` method in [Layer.java](https://github.com/mapsforge/mapsforge/blob/master/mapsforge-map/src/main/java/org/mapsforge/map/layer/Layer.java) but with no implementation. Because all layers (Circle, Polygon etc.) inherit from Layer, it would be useful to be able to register layer listeners and all listeners get notified when a layer has been added. Of course, this can be implemented in Circle, Polygon etc. but this would mean duplicated code. I think this is a basic functionality that Layer should have. Or maybe it would be better to provide methods to add / remove layers in the MapView, e.g.:

```java
public Layers getLayers() {
  return getLayerManager().getLayers();
}
public void addLayerAt(int index, Layer layer) {
  getLayers().add(index, layer);
  for (LayerListener layerListener : layerListeners) {
    layerListener.onLayerAdded(layer, index);
  }
}
public void removeLayerAt(int index) {
  Layer layer = getLayers().get(index);
  getLayers().remove(index);
  for (LayerListener layerListener : layerListeners) {
    layerListener.onLayerRemoved(layer, index);
   }
 }
```

Or did I oversee something and / or layer listeners are already planned?

**Comment by devemux86** [comment, marked as answer]
The methods onAdd / onRemove are used by some derivative layer classes for internal functions.

The layers are added / removed by users / apps, so they already know when they add / remove them.
